multiraft/multiraft.go

```diff
@@ -623,22 +623,22 @@ func (s *state) stop() {
 	s.MultiRaft.Transport.Stop(s.nodeID)
 
 	// Drain the input channels because other threads may be blocking
-	// on these operations.
-	for {
-		select {
-		case op := <-s.createGroupChan:
-			op.ch <- util.Errorf("shutting down")
-		case op := <-s.removeGroupChan:
-			op.ch <- util.Errorf("shutting down")
-		case op := <-s.proposalChan:
-			op.ch <- util.Errorf("shutting down")
-		default:
-			close(s.createGroupChan)
-			close(s.removeGroupChan)
-			close(s.proposalChan)
-			return
+	// on these operations. We can't close these channels unless we
+	// make sure that we never write to them after, but that would mean
+	// an atomic guard before every write access. Better to just deal
```

Comment:
The only time multiraft writes to its own input channels is when it re-submits all pending proposals after an election; an atomic guard there is fine (and much better than deliberately leaking a goroutine to consume from the channels forever).